### PR TITLE
Refined tab autocompletion

### LINKED ISSUE #2788: Better tab complete
It'll be easier to perform nick/item/command completion, so instead of typing the full thing, you just type enough letters then just tab it until you got what you want.

### PR DESCRIPTION
**Pull request by HybridDog**
- Change client side autocompletion
- Add server side completion
- Add core.autocomplete_word helper function
- Use autocompletion for chatcommands
- Add player name autocompletion, minetest.after is used due to its low autocompleting priority

> It would make more sense to do the autocompletion fully client-side.

l don't think so because
autocompletion happens seldom,
no words list or other complex stuff needs to be sent,
stuff is only sent back to the client if necessary,
mods have a lot more possibilities, see https://github.com/minetest/minetest/issues/2788#issuecomment-229966508,
less, or even no, compatibility problems for autocompletion updates are added and
lag doesn't matter because you usually hardly use chat autocompletion when the server lags a lot.

### CODE REVIEW

**Comment by SmallJoker** [comment]
What about server lag? It would be much better to receive the autocomplete list after loading the media on join.

**Comment by HybridDog** [comment]
What do you mean by autocompletion list?

**Comment by SmallJoker** [comment]
With that I actually only meant the available chat commands and custom words that can be autocompleted because the item names and online player are already sent to the client. It would make more sense to do the autocompletion fully client-side, there's no validation check (anticheat) required.

**Comment by HybridDog** [comment]
>It would make more sense to do the autocompletion fully client-side.

l don't think so because
autocompletion happens seldom,
no words list or other complex stuff needs to be sent,
stuff is only sent back to the client if necessary,
mods have a lot more possibilities, see https://github.com/minetest/minetest/issues/2788#issuecomment-229966508 and
less, or even no, compatibility problems for autocompletion updates are added.

**Comment by est31** [comment]
:-1: I do like an autocompletion API but it should be running on the client, not on the server. Yes, its one of the "waiting for clientside scripting" things...

**Comment by HybridDog** [comment]
How about adding both:
In the F10 console the server side autocompletion is used and
in the chat console client side one is performed.

File: src/server.cpp [hunk 1]
```diff
@@ -1654,6 +1654,21 @@ void Server::SendChatMessage(u16 peer_id, const std::wstring &message)
 	}
 }
 
+void Server::SendChatAutocompletion(u16 peer_id, u16 *cursorpos, std::string *message)
```
Comment (est31) [inline src/server.cpp]: Doesn't need to be a pointer, can be a const reference too.

**Comment by est31** [comment]
Redacting my :-1: , we can't block development of minetest just because there is no clientside scripting.

File: src/server.cpp (hunk 1 above)
Comment (HybridDog) [inline src/server.cpp]: fixed

**Comment by nerzhul** [comment]
You can have the auto completion on item names because we send the node list to client at loading. For players we should have a new packet to notify players are joining/leaving to ensure the local playerlist is updated

**Comment by nerzhul** [comment]
This should not be on the server, client only

**Comment by HybridDog** [comment]
>You can have the auto completion on item names because we send the node list to client at loading.

Maybe in future item data is sent on demand that joining works a lot faster, then the list would become sent additionally just for tab autocompletion.

**Comment by HybridDog** [comment]
How about adding both:
In the F10 console the server side autocompletion is used and
in the chat console client side one is performed.

If it's on the server, mods can do much more complex things, e.g. the ambience mod could add a volume adjustment bar similar to the one of $ alsamixer.

**Comment by nerzhul** [comment]
@HybridDog the main problem with that async is to handle it properly before doing UI update, atm UI is synchronous on events, it needs a little bit work

**Comment by HybridDog** [comment]
> the main problem with that async is to handle it properly before doing UI update, atm UI is synchronous on events, it needs a little bit work

Do you mean items being sent on demand? While the items are sent (not every single one in a single packet, a packet containing some, e.g. unknown ones in a new inventory page) the items which aren't received yet are shown as dummies, similar to the unknown items.

**Comment by wilkgr76** [comment]
> If it's on the server, mods can do much more complex things, e.g. the ambience mod could add a volume adjustment bar similar to the one of $ alsamixer.

I would just love a mod that could let me autocomplete the itemstring of the block I'm looking at, seeing how the chat now obscures it.

**Comment by est31** [comment]
Hrmm that's in fact a great idea @HybridDog .

**Comment by nerzhul** [comment]
We don't need autocompletion server side, especially for items as we already have list client side. You modifications client side are great, just add the lua call where you call autocompletion with a callback in complete.

**Comment by HybridDog** [comment]
> You modifications client side are great, just add the lua call where you call autocompletion with a callback in complete.

Do you mean a mixture between client- and server side autocompletion?

**Comment by HybridDog** [comment]
How about enabling the server side autocompletion only for the admin ncurses terminal chat?

**Comment by est31** [comment]
> How about enabling the server side autocompletion only for the terminal_chat_console?

That would be acceptable.

**Comment by HybridDog** [comment]
How can l get the client in terminal_chat_console.cpp?

**Comment by est31** [comment]
`bin/minetest --worldname <your name here> --server --terminal`

**Comment by HybridDog** [comment]
↑
This way you can test the terminal chat console.
l assume there is no client when using the chat console, is there?

**Comment by Megaf** [comment]
I wonder if this would fix https://github.com/minetest/minetest/issues/1934

**Comment by Megaf** [comment]
@nerzhul We do need autocompletion server side, of course! Most of the time I'm only on the server's console.

**Comment by est31** [comment]
@HybridDog no, only server side.

**Comment by HybridDog** [comment]
Megaf, it should fix autocompletion on the terminal chat console. I don't know if it also fixes the client side ingame completion, the function is completely rewritten but the names list, where the possible completions are extracted from, is still the same one.

est31, l assume l could do the completion without sending the informations,
how about adding a privilege which allows the client to use server side autocompletion?

**Comment by Megaf** [comment]
It seems like it needs rebase. And why whasnt this merged yet?

**Comment by HybridDog** [comment]
Megaf, l didn't figure out how to access the cpp_api from the terminal chat console.

**Comment by est31** [comment]
Mhh with CSM around, its better done via a CSM API. cc @red-001

**Comment by Megaf** [comment]
@est31 Can you please enlighten me on what CSM is?

**Comment by est31** [comment]
@Megaf Client side modding, see #5088

**Pull request closed without merging**